upup/pkg/fi/nodeup/nodetasks/issue_cert.go

```diff
@@ -0,0 +1,140 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package nodetasks
+
+import (
+	"bytes"
+	"crypto/x509/pkix"
+	"fmt"
+	"io"
+	"path/filepath"
+
+	"k8s.io/klog"
+	"k8s.io/kops/pkg/pki"
+	"k8s.io/kops/upup/pkg/fi"
+)
+
+type IssueCert struct {
+	Name string
+
+	Signer         string    `json:"signer"`
+	Type           string    `json:"type"`
+	Subject        pkix.Name `json:"subject"`
+	AlternateNames []string  `json:"alternateNames,omitempty"`
+
+	cert *fi.TaskDependentResource
+	key  *fi.TaskDependentResource
+	ca   *fi.TaskDependentResource
+}
+
+var _ fi.Task = &IssueCert{}
+var _ fi.HasName = &IssueCert{}
+
+func (i *IssueCert) GetName() *string {
+	return &i.Name
+}
+
+func (i *IssueCert) SetName(name string) {
+	i.Name = name
+}
+
+// String returns a string representation, implementing the Stringer interface
+func (i *IssueCert) String() string {
+	return fmt.Sprintf("IssueCert: %s", i.Name)
+}
+
+func (i *IssueCert) GetResources() (certResource, keyResource, caResource *fi.TaskDependentResource) {
+	if i.cert == nil {
+		i.cert = &fi.TaskDependentResource{Task: i}
+		i.key = &fi.TaskDependentResource{Task: i}
+		i.ca = &fi.TaskDependentResource{Task: i}
+	}
+	return i.cert, i.key, i.ca
+}
+
+func (i *IssueCert) AddFileTasks(c *fi.ModelBuilderContext, dir string, name string, caName string, owner *string) {
```

Comment:
Not something to address now, but we might want to use a struct just so we can name these fields.  (Or we could have 3 methods that each build one of the file tasks, then it's easy to set e.g. the owner on them.)

Reply:
Most Builders that use certs write them as three files. A few write as two files, omitting the CA. A very few, like the kubelet builder, need to  include them in a configuration file. 

I do want to keep the common case succinct. I'll see what using the struct looks like.